### ISSUE TITLE
Segmentation fault when assigning function to reference in different thread

### ISSUE THREAD

**Issue by radekm** [ocamllabs/ocaml-multicore#13]
The following program

```ocaml
effect Yield : unit

let () =
  let kk = ref (fun () -> failwith "continuation") in
  begin match
    print_endline "0";
    perform Yield;
    print_endline "1";
    Unix.sleep 1;
    perform Yield;
    print_endline "2"
  with
    | () -> print_endline "done"
    | effect Yield k ->
        print_endline "before kk assig";
        kk := (fun () -> continue k ());
        print_endline "after kk assig"
  end;
  print_endline "before spawn";
  Domain.spawn (fun () -> !kk ());
  print_endline "after spawn";
  Unix.sleep 20;
  print_endline "after sleep"
```

always ends with segmentation fault. The output is usually
```
0
before kk assig
after kk assig
before spawn
1
after spawn
before kk assig
Segmentation fault
```

~~It seems that the problem occurs when performing `Yield` after `Unix.sleep 1`.~~ What helps is replacing `Domain.spawn (fun () -> !kk ())` by `!kk ()`.

**Comment by radekm** [comment]
Also note that performing an effect in a different thread doesn't cause the problem. Eg. the following program doesn't end with segmentation fault:

```ocaml

effect Yield : unit

let () =
  begin match perform Yield with
    | () -> ()
    | effect Yield _ -> ()
  end;
  Domain.spawn
    (fun () ->
      match perform Yield with
        | () -> ()
        | effect Yield _ -> print_endline "yield");
  Domain.spawn (fun () -> perform Yield);
  Domain.spawn (fun () -> perform Yield);
  Unix.sleep 1
```

**Comment by radekm** [comment]
The segmentation fault happens when assigning to `kk` - when executing `kk := (fun () -> continue k ())`. But when I assign value which doesn't contain `k` (eg. `(fun () -> continue (failwith "") ())`) then there is no segmentation fault.

**Comment by kayceesrk** [comment]
Fixed by #14.
